sway-lib-std/src/vec.sw

```diff
@@ -0,0 +1,189 @@
+    }
+
+    fn grow(self) {
+        let mut self = self;
+        let new_cap = if self.cap == 0 {
+            1
+        } else {
+            2 * self.cap
+        };
+
+        // Allocate for `new_cap` elements.
+        asm(size: new_cap * __size_of::<T>()) {
+            aloc size;
+        };
+        let new_ptr = heap_ptr() + 1;
+
+        // Copy old contents into newly-allocated memory.
+        asm(new_ptr: new_ptr, old_ptr: self.ptr, size: self.cap * __size_of::<T>()) {
+            mcp new_ptr old_ptr size;
+        };
+
+        self.ptr = new_ptr;
+        self.cap = new_cap;
+    }
+}
+
+/// A contiguous growable array type, written as `Vec<T>`, short for 'vector'.
+pub struct Vec<T> {
+    buf: RawVec,
+    len: u64,
+}
+
+impl<T> Vec<T> {
+    /// Constructs a new, empty `Vec<T>`.
+    ///
+    /// The vector will not allocate until elements are pushed onto it.
+    pub fn new() -> Self {
+        Vec {
+            buf: ~RawVec::new(),
+            len: 0,
+        }
+    }
+
+    /// Constructs a new, empty `Vec<T>` with the specified capacity.
+    ///
+    /// The vector will be able to hold exactly `capacity` elements without
+    /// reallocating. If `capacity` is 0, the vector will not allocate.
+    ///
+    /// It is important to note that although the returned vector has the
+    /// *capacity* specified, the vector will have a zero *length*.
+    pub fn with_capacity(capacity: u64) -> Self {
+        Vec {
+            buf: ~RawVec::with_capacity(capacity),
+            len: 0,
+        }
+    }
+
+    /// Appends an element to the back of a collection.
+    pub fn push(self, value: T) {
+        let mut self = self;
+        // If there is insufficient capacity, grow the buffer.
+        if self.len == self.buf.capacity() {
+            self.buf.grow();
+        };
+        // Get a pointer to the end of the buffer, where the new element will
+        // be inserted.
+        let end = self.buf.ptr() + self.len * __size_of::<T>();
+
+        if !__is_reference_type::<T>() {
+            // If `T` is not a reference type, then it is a one-word primitive.
+            // Simply store it at `end` pointer.
+            asm(end: end, value: value) {
+                sw end value i0;
+            };
+        } else {
+            // If `T` is a reference type, then it points to a memory range.
+            // Copy the `value`'s memory range to the range pointed to by `end`.
+            let size = __size_of::<T>();
+            asm(end: end, ptr: value, size: size) {
+                mcp end ptr size;
+            };
+        };
+
+        // Increment length.
+        self.len = self.len + 1;
+    }
+
+    /// Gets the capacity of the allocation.
+    pub fn capacity(self) -> u64 {
+        self.buf.cap
+    }
+
+    /// Clears the vector, removing all values.
+    ///
+    /// Note that this method has no effect on the allocated capacity
+    /// of the vector.
+    pub fn clear(self) {
+        let mut self = self;
```

Comment:
Why is line 140 needed?

Reply:
It was failing to build because it was complaining that self wasn't mutable. I tried adding `mut` in front of `self` in the parameters but it was giving me a type error, so I did this as a hack to get it to build.

Reply:
I mean, I don't understand how assigning self back to self does anything. What is the point of this line?

Reply:
Sorry, I mistyped! Putting mut in front of self in the parameters was a _parse_ error. So basically all this line is doing is reassigning self to self, but with the mut keyword, so that it can avoid the parse error.

Reply:
But why do you _need_ to assign self back to self? Isn't that basically a no-op?

Reply:
Yes, and no. Normally, in Sway, and in Rust, you would add the `mut` keyword in front of `self` in the function parameters of `clear`. However doing that on `master` right now is either not supported or there is a bug, and doing that generates a parse error. So, this _is_ a no-op in that there are no changes to self, but it's also not a no-op because it allows the compiler to see that `self` is mutable, which is needs to do, in order to assign to it in the `clear` function.

I don't think this code should be merged! We should address the parse error first. I just wanted to demonstrate that this library does indeed compile on `master`, so I threw this hacky line in.

Reply:
Relevant: https://github.com/FuelLabs/sway/issues/1188